### PR TITLE
CNS-368: introduce TimerStore to manage future timeouts

### PR DESCRIPTION
**Pull request by orenl-lava**
TimerStore manages timers to efficiently support future timeouts. Timeouts can be based on either block-height or block-timestamp. When a timeout occurs, a designated callback function is called with the details (ctx and data). 

Note: this PR depends on #403.

### CODE REVIEW

File: common/timer.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+//
+// 3. TimerStore tracks the next-timeout for both block-height/block-timestamp. On
+// every call to Tick(), it tests the current ctx's block height/timestamp against the
+// respective next-timeout:
+//
+// 4. If the next-timeout is reached/passed, then it will iterate through the timer
+// entries and invoke the (respective) callback for those entries; And finally it will
+// advance the (respective) next-timeout.
+
+// Thus, iterating on the prefix "packages_Entry_Index_" would yield all the package
```
Comment (oren-lava) [inline common/timer.go:50]: From here to the example, I'm not sure how this is related to the Timer explanation. Also, you have two "4."

File: common/timer.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+//     prefix: packages_Timer_Next_       key: BlockHeight      value: 150
+//     prefix: packages_Timer_Next_       key: BLockTimer       value: MaxInt64
+//     prefix: packages_Timer_Value_          key: 150          data: "first"
+//     prefix: packages_Entry_Value_          key: 180          data: "second"
+//
+// 3. TimerStore tracks the next-timeout for both block-height/block-timestamp. On
+// every call to Tick(), it tests the current ctx's block height/timestamp against the
+// respective next-timeout:
+//
+// 4. If the next-timeout is reached/passed, then it will iterate through the timer
```
Comment (oren-lava) [inline common/timer.go:46]: A thought: isn't it problematic that the callbacks might be invoked long after they were supposed to be called?
For example, say I have a callback that is supposed to be called in block 100 and the user ticks the timer only on block 1000.
Maybe there should be a global timer that is incremented on every new block so there will be a single reference of the current time to all timers

**Review by oren-lava — DISMISSED** [review]

File: common/timer.go (hunk 2 above)
Comment (orenl-lava) [reply common/timer.go:46]: The TimerStore is like a library, and users may use it as they see fit.
If exact timeouts are needed, the user can call AdvanceBlock() on every BeginBlock() of its own module.
If it is sufficient to have callbacks on epoch start/end, then the user can call AdvanceBlock() on epoch start/end (and save some CPU not checking blocks in between).
As long as the callback timing is deterministic, and the user controls the "AdvanceBlock()" calls, things are fine.

Comment (orenl-lava) [reply common/timer.go:46]: Ideally, TimerStore will be its own module, and will have its own BeginBlock(), and won't depend on the user code calling AdvanceBlock(). There can also be a parameter to set the frequency of checking. Currently left for future work.

File: common/timer.go (hunk 1 above)
Comment (orenl-lava) [reply common/timer.go:50]: Sigh. Self copyright infringement :)
Will clean up.

File: common/timer.go (hunk 2 above)
Comment (oren-lava) [reply common/timer.go:46]: I think your comment about calling AdvanceBlock() on every BeginBlock() of the user's module for exact timeouts should be explicitly documented in the TimerStore documentation as a best/common practice (so a forgetful user, like me, won't forget to do it)

File: common/types/keys.go [hunk 3]
```diff
@@ -6,13 +6,45 @@ func KeyPrefix(p string) []byte {
...
+)
+
+var NextTimerKey = []string{
+	"NextBlock", // for BlockHeight
+	"NextDate",  // for BlockTime
+}
+
+var TimerTypePrefix = []string{
+	"Block_", // for BlockHeight
+	"Date_",  // for BlockTime
```
Comment (Yaroms) [inline common/types/keys.go:37]: lets stick with time instead of date?

**Review by Yaroms — CHANGES_REQUESTED** [review]

**Review by omerlavanet — CHANGES_REQUESTED** [review]
very nice work! high quality code

File: common/timer.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+func (tstore *TimerStore) setNextTimeout(ctx sdk.Context, which types.TimerType, value uint64) {
+	store := tstore.getStore(ctx, types.NextTimerPrefix)
+	b := types.EncodeKey(value)
+	store.Set([]byte(types.NextTimerKey[which]), b)
+}
+
+func (tstore *TimerStore) addTimer(ctx sdk.Context, which types.TimerType, value uint64, data string) {
+	store := tstore.getStoreTimer(ctx, which)
+	store.Set(types.EncodeKey(value), []byte(data))
```
[the remark this answers is not in the document]
Comment (orenl-lava) [reply common/timer.go:136]: See the "How does it work" section (bullet 2) at the beginning of the file.

File: common/timer_test.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		{op: "addheight", name: "add timer no 3", value: 140, data: "no-3."},
+		{op: "addheight", name: "add timer no 4", value: 150, data: "no-4."},
+		{op: "tickheight", name: "tick before all", value: 110, fire: 0},
+		{op: "tickheight", name: "tick between no-2, no-3", value: 135, fire: 2, data: "no-1.no-2."},
+		{op: "nextheight", name: "next timeout no-3", value: 140},
+		{op: "tickheight", name: "tick after all", value: 155, fire: 2, data: "no-3.no-4."},
+	}
+
+	testWithTimerTemplate(t, playbook, 1)
+}
```
[the remark this answers is not in the document]
Comment (orenl-lava) [reply common/timer_test.go:164]: that's guaranteed by cosmos-sdk API KVStorePrefixIterator.
(see https://github.com/cosmos/cosmos-sdk/blob/main/store/types/utils.go)

**Review by Yaroms — APPROVED** [review]

**Pull request merged**